code/modules/arcade/mob_hunt/mob_datums.dm

```diff
@@ -0,0 +1,475 @@
+			break
+		//clear and generate a fresh list of turfs in the selected area, weighted based on white/black lists
+		var/list/possible_turfs = get_possible_turfs(spawn_area)
+		if(!possible_turfs.len)		//If we don't have any possible turfs, this attempt was a failure. Try again.
+			continue
+		//if we got this far, we're spawning on this attempt, hooray!
+		spawn_point = pickweight(possible_turfs)
+		break
+	if(!spawn_point)
+	//if we get to this, we failed every attempt to find a suitable turf for EVERY area in our list of possible areas. DAMN.
+		log_admin("No acceptable turfs to spawn [type] on could be located. Possible code/mapping error, or someone replaced/destroyed all the acceptable turf types?")
+		return 0
+	return 1
+
+/datum/mob_hunt/proc/get_possible_areas()
+	var/list/possible_areas = list()
+	//setup, sets all station areas (and subtypes) to weight 1
+	for(var/A in the_station_areas)
+		if(A == /area/holodeck)	//don't allow holodeck areas as possible spawns since it will allow it to spawn in the holodeck rooms on z2 as well
+			continue
+		if(A in possible_areas)
+			continue
+		for(var/areapath in typesof(A))
+			possible_areas[areapath] = 1
+	//primary type preferences
+	if(primary_type)
+		for(var/A in primary_type.area_whitelist)
+			for(var/areapath in typesof(A))
+				possible_areas[areapath] += 4
+		for(var/A in primary_type.area_blacklist)
+			for(var/areapath in typesof(A))
+				possible_areas[areapath] -= 2
+	//secondary type preferences
+	if(secondary_type)
+		for(var/A in secondary_type.area_whitelist)
+			for(var/areapath in typesof(A))
+				possible_areas[areapath] += 4
+		for(var/A in secondary_type.area_blacklist)
+			for(var/areapath in typesof(A))
+				possible_areas[areapath] -= 2
+	//mob preferences
+	for(var/A in area_whitelist)
+		for(var/areapath in typesof(A))
+			possible_areas[areapath] += 4
+	for(var/A in area_blacklist)
+		for(var/areapath in typesof(A))
+			possible_areas[areapath] -= 2
+	//removes "bad areas" which shouldn't be on-station but are subtypes of station areas. probably should the unused ones and consider repathing the rest
+	var/list/bad_areas = list(subtypesof(/area/construction), /area/solar/derelict_starboard, /area/solar/derelict_aft, /area/solar/constructionsite)
+	for(var/A in bad_areas)
+		possible_areas -= A
+	//weight check, remove negative or zero weight areas from the list, then return the list.
+	for(var/areapath in possible_areas)
+		//remove any areas that shouldn't be on the station-level
+		if(possible_areas[areapath] < 1)
+			possible_areas -= areapath
+			continue
+	return possible_areas
+
+/datum/mob_hunt/proc/get_possible_turfs(area/spawn_area)
+	if(!spawn_area)
+		return list()
+	var/list/possible_turfs = list()
+	//setup, sets all turfs in spawn_area to weight 1
+	for(var/turf/T in spawn_area)
+		if(T.z > 1)		//mobs will only consider station-level turfs for spawning. Largely here so we won't have to worry about mapping errors or mobs on the derelict solars
```

Comment:
This isn't how we check for the station's Z-level(s) anymore. `is_station_level(z)`.